microshift_running_apps/microshift-operators-olm.adoc

```diff
@@ -0,0 +1,62 @@
+:_mod-docs-content-type: ASSEMBLY
+[id="olm-with-microshift"]
+= Using Operator Lifecycle Manager with {microshift-short}
+include::_attributes/attributes-microshift.adoc[]
+:context: microshift-operators-olm
+
+toc::[]
+
+OLM is used in {microshift-short} for installing and running optional link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html/architecture/control-plane#olm-operators_control-plane[add-on Operators].
+
+* Cluster Operators as applied in {ocp} are not used in {microshift-short}.
+* You can add your own catalogs for the Operators you want to use with your applications.
+* You must use the CLI to conduct OLM activities with {microshift-short}. The console and OperatorHub GUIs are not available, but you can take all the same actions with the CLI.
+* Install the link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html/cli_tools/opm-cli#cli-opm-install[`opm` CLI]
+
+[IMPORTANT]
+====
+Before using an Operator, verify with the provider whether the Operator is supported on {product-title}.
+====
+
+[id="microshift-file-based-olm-catalogs"]
+== File-based Operator catalogs
+
+To run {microshift-short} with your own curated catalogs and OLM, use the file-based catalog structure. For a complete explanation of the file-based structure, see the "Additional resources" section.
```

Comment:
While it doesn't ship with it, it is definitely true that in order to have a global catalog on the cluster the user will have to create the openshift-marketplace namespace and put their catalog source there. See:

https://github.com/openshift/microshift/blob/main/assets/optional/operator-lifecycle-manager/0000_50_olm_08-catalog-operator.deployment.yaml#L54

Maybe that's something we want to revisit? We'd have to modify that field and then tell the user to create catalogs in a different namespace and ship the openshift-marketplace namespace with nothing in it by default.

Reply:
Yeah, correct. `microshift-olm` [creates](https://github.com/openshift/microshift/blob/main/assets/optional/operator-lifecycle-manager/0000_50_olm_00-namespace.yaml) two namespaces: `openshift-operator-lifecycle-manager` and `openshift-operators`.
In tests we're using the `openshift-operators` for the [Catalog](https://github.com/openshift/microshift/blob/main/test/assets/olm/catalog-source.yaml#L5) and [as target namespace for the cert-manager](https://github.com/openshift/microshift/blob/main/test/assets/olm/subscription-cert-manager.yaml#L5) but I'm not sure now if this is as intended?

Reply:
@pmtk you can, it just means that that catalog can only be used for operator installations in that openshift-operators namespace. If you want to install an operator in a different namespace, you have to run the catalog again in whatever namespace you're installing the operator in.

The only exception to this is if you create the catalog in the global catalog namespace. For openshift, based on that variable we pass into the olm manifest, that's the `openshift-marketplace` namespace.

Keep in mind that installing every operator in one namespace isn't really an option because operators in OLM have a concept of watch scope -- for example, some operators only support watching their own namespace, others support watching every namespace in the cluster -- and all operators installed in a given namespace must have the same scope.

Reply:
I have little experience using OLM, not even mentioning configuration or implementation details, so here we need to rely on OLM team knowledge.

@kevinrizza if you could guide us on changes we should make to ship best sane defaults, then I can make that change so they will make it in time to be part of 4.15.0.

Reply:
forcing all operators to run in openshift-operators is not a good idea, as kevin mentioned.
I would prefer the catalog to run in a global namespace, and would prefer "openshift-operator-lifecycle-manager" to be used for that (to me, it feels like it belongs there). @kevinrizza , what do we need to do to change the olm config in regards to that?

Reply:
got it. then lets discuss internally in microshift team before doing something too quickly here.

Reply:
> Imo it would be simpler and not much more effort to just ship an empty openshift-marketplace namespace so that it conforms to how we ship the rest of OpenShift

agree with it.

Reply:
Some adjustments made here: https://70128--ocpdocs-pr.netlify.app/microshift/latest/microshift_running_apps/microshift-operators-olm#olm-creating-catalog-from-index_microshift-operators-olm others tbd from discussion

Reply:
for https://70128--ocpdocs-pr.netlify.app/microshift/latest/microshift_running_apps/microshift-operators-olm#olm-creating-catalog-from-index_microshift-operators-olm, please remove `d Check the package manifest:` because there is no packagemanifest CR in microshift (ocp has it)

Reply:
for `If you want the catalog source to be available globally in all namespaces, specify the openshift-marketplace namespace. Use this if you want to deploy different operators to different namespaces.`, it is totally correct.

for `Otherwise, you can specify another namespace for the catalog (e.g. "openshift-operators") to be scoped and available only for that namespace. Then all operators deployed using this catalog have to run in the same namespace.`, there is one reminder. if the openshift-operators is used, the operator installed in this ns is AllNs mode because the og in this ns is default all ns mode. it means the all ns can see it with copied csv.
so, if the catalog can not be used for all ns, and operator also can not be used for all ns, better to create new ns except for openshift-operators.  if the catalog can not be used for all ns, and operator be used for all ns, could use  openshift-operators.

Reply:
more information:
there are two kinds of things which is different:
1, where to install catalog so that where we use the catalog
2, where to install operator so that where the operator watch.

for 1, if the catalog is installed openshift-marketplace, we can use the catalog in any ns.
          if the catalog is installed in other ns (not openshift-marketplace), we can only use it in same ns.

for 2, it depends on install mode by operator group CR.
          if you install operator in openshif-operators ns, there is default og with all ns mode. the operator installed in openshif-operators will watch all ns.
          if you install operator in other ns which has no default og, you could create og with single ns mode or own ns mode. the operator installed in this ns will only watch this ns.

it seems mix these two kinds of thing. so, xhttps://70128--ocpdocs-pr.netlify.app/microshift/latest/microshift_running_apps/microshift-operators-olm#microshift-deploy-operators_microshift-operators-olm confuse me. 

for me, 
if customer want to use one catalog in different ns, it have to be in openshif-marketplance ns. and then if the customer want to install operator to watch all ns, the operator should be installed on openshift-operators ns referening this catalog. if the customer want to install operator to watch its own ns, the customer can create one ns and create og with own ns mode in this ns and install operator in this ns refering this catalog.

if customer does not want to use the catalog for different ns, customer could create one ns to install catalog. and then customer create og and install operator in this ns refering the catalog in this ns. if og is all ns mode, the operator will watch all ns. if og is own ns mode, the operator will only watch this ns.
(of course, you could use openshif-operators ns not creating new ns, please NOTE: and then you have to install operator in this ns, and the operator installed have to watch all ns. and it required operator support all ns mode. or else, the installation will fail)

so, we'd better not mix catalog creation and operator creation in openshif-operators because I worry about that the operator customer want to install does not support all ns mode, and also worry about that the customer does not want operator watch all ns.

of course, if the custer want to install opeator with all ns mode supported and watch all ns, it is ok to install both catalog and operator in openshift-operators ns.

Reply:
@kuiwang02 Do you know of any single page in the OLM docs that explains all of this together? I don't need to write something we already have, I just haven't found catalog creation details in the OLM docs. Am I missing it? In any case, I separated the "install an operator" examples, and added some overview bullets to the concept section on creating catalogs.

Reply:
@ShaunaDiaz I am sorry I do not know if there is such olm doc. it is my understanding after learning olm.